Improve math equation typography.

MarkCallow opened the pull request:
Fixes #226.

lexaknyazev reviewed (changes requested):
Please also include the `KTXwriterScParams` revision history typo fix in this PR.

MarkCallow commented:
https://github.com/KhronosGroup/KTX-Specification/pull/227/commits/76045a4ea7ae7da8fb038fa463354975ca88be9c addresses most of these review issues. I will close those it addresses.

MarkCallow commented on ktxspec.adoc:
```diff
@@ -868,30 +871,30 @@ faces, all rows (or rows of blocks) and all pixels (or blocks) in
 The total size of the image data from
-stem:[levels[numLevels-1\].byteOffset] (i.e., after the first
-`mipPadding`, if any) until the end of the file is:
+stem:[\texttt{levels}[\textit{num_levels}-1\].\texttt{byteOffset}] (i.e.,
```
I'm not sure exactly what you want here. Please elucidate.

lexaknyazev replied:
`stem:[\texttt{levels[}\textit{num_levels}-1\texttt{\].byteOffset}]`

MarkCallow commented on ktxspec.adoc:
```diff
@@ -868,30 +871,30 @@ faces, all rows (or rows of blocks) and all pixels (or blocks) in
-\sum_{p=1}^{numLevels-1}
-\left\lceil{\frac{levels[p].byteLength}{requiredAlignment}}\right\rceil
-\times requiredAlignment
+\texttt{levels}[0].\texttt{byteLength} +
+\sum_{p=1}^{\textit{num_levels}-1}
+\left\lceil{\frac{\texttt{levels}[p].\texttt{byteLength}}{\textit{required_alignment}}}\right\rceil
+\times \textit{required_alignment}
```
I'm not sure exactly what you want here. Please elucidate.

lexaknyazev replied:
```
\texttt{levels[0].byteLength} +
\sum_{p=1}^{\textit{num_levels}-1}
\left\lceil{\frac{\texttt{levels[}p\texttt{].byteLength}}{\textit{required_alignment}}}\right\rceil
\times \textit{required_alignment}
```

MarkCallow commented on ktxspec.adoc:
```diff
@@ -1445,10 +1448,10 @@ When `<<_supercompressionscheme,supercompressionScheme>> != 0` these
+only required when `<<_supercompressionscheme,supercompressionScheme>> == 0`.
 
 _Texel block size_ is as given for the `vkFormat` value in
 {url-khr-vulkan}/specs/1.2-extensions/html/vkspec.html#formats-compatibility[section
```
If I do that I should update all the Vulkan references to 1.4. Is that okay?

lexaknyazev replied:
We should use at least Vulkan 1.3 because ASTC HDR, forbidden formats from `VK_EXT_ycbcr_2plane_444_formats`, and formats from `VK_EXT_4444_formats` were added to Vulkan 1.3 core and the KTX spec refers to them without `_KHR` or `_EXT` suffixes.

Additionally we could update enums for `VK_KHR_maintenance5` formats because they were added to Vulkan 1.4 core.

All of this should probably be done in a separate PR.

lexaknyazev commented on ktxspec.adoc:
```diff
@@ -596,22 +600,23 @@ to signal which faces are present.
-levelCount - 1] and stem:[base = 0].  stem:[level_p] must not be
-greater than the maximum possible, stem:[level_{max}], where
+the level with the smallest size images, stem:[\textit{level}_p]
+to that with the largest size images, stem:[\textit{level}_{base}]
```
`base` needs wrapping too.

lexaknyazev commented on ktxspec.adoc:
```diff
@@ -849,8 +854,9 @@ value must be 0.
 mip level within the <<_mip_level_array,_Mip Level Array_>> The array is ordered
-starting with stem:[level_{base}] (the level with the largest size images)
-at index _0_. Image for stem:[level_p] will be found at index _p_.
+starting with stem:[\textit{level}_{base}] (the level with the largest size images)
```
`base` needs wrapping too.

MarkCallow replied:
Fixed in https://github.com/KhronosGroup/KTX-Specification/pull/227/commits/f37f20c3bdd5a13b8b6dedb75efb29ebfceaafd3.

lexaknyazev commented on ktxspec.adoc:
```diff
@@ -849,8 +854,9 @@ value must be 0.
-starting with stem:[level_{base}] (the level with the largest size images)
-at index _0_. Image for stem:[level_p] will be found at index _p_.
+starting with stem:[\textit{level}_{base}] (the level with the largest size images)
+at index stem:[0]. The image for stem:[|textit{level}_p] will be found at index
```
Should be `\` instead of `|`.

MarkCallow replied:
Will make a separate PR for the Vulkan references.

I disagree with this change

```
stem:[\texttt{levels[}\textit{num_levels}-1\texttt{\].byteOffset}]
```

and similar. I think the current way with the skinny bracket looks better. The first "levels ..." on the first line in this image is with your suggested change. The second is as it is now.

<img width="918" alt="Screenshot 2025-02-17 at 16 41 10" src="https://github.com/user-attachments/assets/36cc0596-f30b-4dae-a58b-d80d46482d76" />

The other issues are resolved in https://github.com/KhronosGroup/KTX-Specification/pull/227/commits/f37f20c3bdd5a13b8b6dedb75efb29ebfceaafd3.

lexaknyazev replied:
> I think the current way with the skinny bracket looks better.

That's fair. My main concern is the extra gap after the dot. How about this:
```
stem:[\texttt{levels}[\textit{num_levels}-1\]\texttt{.byteOffset}]
```

MarkCallow replied:
It doesn't seem to make much difference to the gap after the dot. It makes the spaces on both sides of the dot equal though. Is that what you want? The first expression has the change. The second is the original.

<img width="786" alt="Screenshot 2025-02-17 at 21 42 15" src="https://github.com/user-attachments/assets/42a59d7b-aab2-44bf-a13e-5acf40a7f136" />

MarkCallow replied:
Regarding the Vulkan references, the registry main page now has only a single link to "latest" which points to a doc with the core spec. + extensions. I think I should remove the version number from the Vulkan references and also change the `vkFormat` section which has separate links to both core and extension. Both 1.2 links are redirected to "latest" by the Vulkan registry.

Off topic for this PR but I want to agree direction before I make a PR to update the links.

MarkCallow replied:
Maybe I should link to https://docs.vulkan.org/spec/latest/chapters/formats.html for the reference in the `vkFormat` section. What do you think? I think the main Vulkan reference should be to the single html spec. document in the registry.

lexaknyazev replied:
Agreed on using the new chapter link for `VkFormat` and the "latest" for everything else.

MarkCallow replied:
I just updated to Asciidoctor 2.0.23. I was previously using 2.0.20. I see very little difference. I even put copies of old and new on separate lines to see if that changed anything.

<img width="822" alt="Screenshot 2025-02-17 at 22 32 48" src="https://github.com/user-attachments/assets/46aa29da-a98d-4049-8e09-bf64cc4cafbf" />

Your rendering is different in a number of respects.  I wonder why.

I don't want to spend any more time on this so I will make the change.

MarkCallow replied:
Fixed in https://github.com/KhronosGroup/KTX-Specification/pull/227/commits/4f7737a7e4b6b74b35a298772f5ed5dd80fe4817.

MarkCallow commented on ktxspec.adoc:
```diff
@@ -863,40 +869,40 @@ for mip level _p_. It is the offset of the first byte after any
-\sum_{p=1}^{numLevels-1}
-\left\lceil{\frac{levels[p].byteLength}{requiredAlignment}}\right\rceil
-\times requiredAlignment
+\texttt{levels}[0].\texttt{byteLength} +
```
Fixed in https://github.com/KhronosGroup/KTX-Specification/pull/227/commits/f5f61b85968c7739adb0c8081b14997702043b84.

lexaknyazev reviewed (approved):
I think this PR is good to go but there is a new issue.

Apparently, all `\textit` blocks are rendered incorrectly (using the Times font) with the default "HTML-CSS" MathJax renderer in Chrome. This issue is not present in other browsers. Switching the renderer to "Common HTML" (right-click on any math expression) also bypasses the issue.

I'll take a look.

lexaknyazev commented:
The "HTML-CSS" MathJax renderer replaces regular letters used outside of `\textit` blocks with special Unicode math characters. For example, `p` (0x70) is replaced with `𝑝` (0x1D45D). Then all such symbols (regular and special) use the `font-family: STIXGeneral-Italic` style in Chrome and Safari.

For some reason, Chrome (at least on macOS) does not recognize the `STIXGeneral-Italic` font family and the following happens:
- Special math symbols fall back to `STIX Two Math` and the issue remains hidden.
- Regular symbols fall back to `Times` (or whatever the default font face is) and the issue appears.

Safari, on the other hand, does recognize `STIXGeneral-Italic` font family and internally replaces it with `Times-Italic` for regular symbols thus hiding the issue completely.

The proper solution is to switch to the "CommonHTML" MathJax renderer, which is platform-independent. The simplest way to do that is to post-process the generated HTML and replace `TeX-MML-AM_HTMLorMML` with `TeX-MML-AM_CHTML`. The existing `inline-images.rb` script could be updated to perform this change like this:

```diff
diff --git a/inline-images.rb b/inline-images.rb
index cb54f67..130bd0e 100644
--- a/inline-images.rb
+++ b/inline-images.rb
@@ -6,6 +6,7 @@ require 'base64'
 Asciidoctor::Extensions.register do
   postprocessor do
     process do |document, output|
+      output.sub!('TeX-MML-AM_HTMLorMML', 'TeX-MML-AM_CHTML') if document.basebackend? 'html'
       _replace_urls(output) if document.basebackend? 'html'
     end
   end
```

MarkCallow commented:
Thanks for the investigation into the MathJax issue. I've added your suggested workaround. Please file a bug with Chrome.

lexaknyazev commented:
This seems to be a macOS "feature" rather than a Chrome bug. The "STIXGeneral" font was supported up until macOS 12 but then it got demoted to "Document-support" starting with macOS 13. In the latter case, fonts seem to become less accessible to applications. For reference:
- https://support.apple.com/en-us/103203
- https://support.apple.com/en-us/103197

MarkCallow commented:
> This seems to be a macOS "feature" rather than a Chrome bug. The "STIXGeneral" font was supported up until macOS 12 but then it got demoted to "Document-support" starting with macOS 13. In the latter case, fonts seem to become less accessible to applications.

What struck me as a bug and led to my comment was replacing a missing italic font with a non-italic font when "STIXGeneral" is not recognized. Did I misunderstand your original report?

Unrelated, is there a way in Asciidoctor to tell it to use "CommonHTML" in the Mathjax it generates?

lexaknyazev commented:
> replacing a missing italic font with a non-italic font when "STIXGeneral" is not recognized

That behavior does not seem to be specified anywhere so browsers do as they wish. I think the issue is mainly caused by directly using `STIXGeneral-Italic` as a font name. If it was just `STIXGeneral` with `font-style: italic`, the fallback formatting would work much better.

> Unrelated, is there a way in Asciidoctor to tell it to use "CommonHTML" in the Mathjax it generates?

No, it's hard-coded. Also the MathJax version bundled with Asciidoctor is quite old.
